### PR TITLE
Throw DaprException for errors in DaprClient

### PR DESCRIPTION
**Pull request by rynowak**
# Description

This changes DaprClient to throw DaprException (or a subclass), wrapping the root cause, in locations where DaprClient does networking or handles user data. This allows for simpler error handling, more topical error messages, and in the case of invoke: more information is available.

There's also work here to unify the two classes we had called `DaprException`. The new `DaprException` class is pretty plain, and used for the base class of other exception types.

## Issue reference

#516 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: src/Dapr.Actors/Client/ActorProxy.cs [hunk 1]
```diff
@@ -169,7 +169,7 @@ public Task InvokeMethodAsync(string method, CancellationToken cancellationToken
             this.ActorId = actorId;
             this.ActorType = actorType;
             this.ActorMessageBodyFactory = client.GetRemotingMessageBodyFactory();
-            this.JsonSerializerOptions = options?.JsonSerializerOptions ?? this.JsonSerializerOptions;
+            this.JsonSerializerOptions = options?.JsonSerializerOptions ?? new JsonSerializerOptions(JsonSerializerDefaults.Web);
```
Comment (rynowak) [inline src/Dapr.Actors/Client/ActorProxy.cs:172]: See comments on the commit about why this change was needed.

File: src/Dapr.Actors/DaprHttpInteractor.cs [hunk 2]
```diff
@@ -373,25 +373,25 @@ HttpRequestMessage FinalRequestFunc()
                 }
                 catch (Exception ex)
                 {
-                    throw new DaprException(string.Format("ServerErrorNoMeaningFulResponse", response.StatusCode), ex);
+                    throw new DaprApiException(string.Format("ServerErrorNoMeaningFulResponse", response.StatusCode), ex);
```
Comment (rynowak) [inline src/Dapr.Actors/DaprHttpInteractor.cs:376]: Actors kinda has its own style of errors based on the HTTP interface. Some of the code changed, the functionality should be the same.

File: src/Dapr.Client/DaprException.cs [hunk 3]
```diff
@@ -31,5 +33,15 @@ public DaprException(string message, Exception innerException)
             : base(message, innerException)
         {
         }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DaprException"/> class with a specified context.
+        /// </summary>
+        /// <param name="info">The <see cref="SerializationInfo" /> object that contains serialized object data of the exception being thrown.</param>
+        /// <param name="context">The <see cref="StreamingContext" /> object that contains contextual information about the source or destination. The context parameter is reserved for future use and can be null.</param>
+        protected DaprException(SerializationInfo info, StreamingContext context)
```
Comment (rynowak) [inline src/Dapr.Client/DaprException.cs:42]: Serialization constructor is needed because the actors code has it.

File: test/Dapr.Client.Test/PublishEventApiTest.cs [hunk 4]
```diff
@@ -109,11 +111,10 @@ public async Task PublishEventAsync_CanPublishTopicWithNoContent_WithMetadata()
             var task = daprClient.PublishEventAsync(TestPubsubName, "test", metadata);
             httpClient.Requests.TryDequeue(out var entry).Should().BeTrue();
             var request = await GrpcUtils.GetRequestFromRequestMessageAsync<PublishEventRequest>(entry.Request);
-            var jsonFromRequest = request.Data.ToStringUtf8();
 
             request.PubsubName.Should().Be(TestPubsubName);
             request.Topic.Should().Be("test");
-            jsonFromRequest.Should().Be("\"\"");
+            request.Data.Length.Should().Be(0);
```
Comment (rynowak) [inline test/Dapr.Client.Test/PublishEventApiTest.cs:117]: This was a bug identified on inspecting the publish code. Publishing "no content" would result in `""` whereas providing `null` would result in "no content".

File: test/Dapr.Client.Test/PublishEventApiTest.cs [hunk 5]
```diff
@@ -87,7 +89,7 @@ public async Task PublishEventAsync_CanPublishTopicWithNoContent()
 
             request.PubsubName.Should().Be(TestPubsubName);
             request.Topic.Should().Be("test");
-            jsonFromRequest.Should().Be("\"\"");
```
Comment (vinayada1) [inline test/Dapr.Client.Test/PublishEventApiTest.cs:90]: Do we need a UT for WithContent_ContentEqualsNull?

File: src/Dapr.Client/DaprClientGrpc.cs [hunk 6]
```diff
@@ -234,24 +267,100 @@ public override Task<HttpResponseMessage> InvokeMethodWithResponseAsync(HttpRequ
...
+                return await this.httpClient.SendAsync(request, cancellationToken);
+            }
+            catch (HttpRequestException ex)
+            {
+                // Our code path for creating requests places these keys in the request properties. We don't want to fail
+                // if they are not present.
+                request.Properties.TryGetValue(AppIdKey, out var appId);
+                request.Properties.TryGetValue(MethodNameKey, out var methodName);
+
+                throw new InvocationException(
```
Comment (vinayada1) [inline src/Dapr.Client/DaprClientGrpc.cs:281]: question - why do we have a special InvocationException class here whereas everything else uses DaprException?

File: src/Dapr.Client/DaprApiException.cs [hunk 7]
```diff
@@ -12,79 +12,79 @@ namespace Dapr.Actors
...
         /// <summary>
-        /// Initializes a new instance of the <see cref="DaprException"/> class with a specified error
+        /// Initializes a new instance of the <see cref="DaprApiException"/> class with a specified error
         /// message and a reference to the inner exception that is the cause of this exception.
         /// </summary>
         /// <param name="message">The error message that explains the reason for the exception.</param>
         /// <param name="innerException">The exception that is the cause of the current exception.</param>
-        public DaprException(string message, Exception innerException)
-            : this(message, innerException, Constants.Unknown, false)
+        public DaprApiException(string message, Exception innerException)
```
Comment (vinayada1) [inline src/Dapr.Client/DaprApiException.cs:40]: Is DaprApiException like DaprActorException...the naming isn't very clear

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 6 above)
Comment (rynowak) [reply src/Dapr.Client/DaprClientGrpc.cs:281]: This is a subclass of DaprException - so a catch for DapException would handle this as well. The reason for the subclass is that we have more information and context to include for invocation. Also needing to handle errors in this case is *much* more common since invoke calls business logic on the side - error cases are **expected**.

Basically, it makes sense to try and optimize this experience. The errors in the other cases are fairly generic, and could be improved in the future as needed based on feedback (without breaking changes).

File: test/Dapr.Client.Test/PublishEventApiTest.cs (hunk 5 above)
Comment (rynowak) [reply test/Dapr.Client.Test/PublishEventApiTest.cs:90]: If you mean a for passing in `null` as a value, I don't mind having one. The thing is .. this isn't testing what our code does, its mostly testing what our code DOESN'T do 😆 I don't mind having one test, but I don't want to go on an extensive journey of testing the underlying behaviors of the JSON serializer.

File: src/Dapr.Client/DaprApiException.cs (hunk 7 above)
Comment (rynowak) [reply src/Dapr.Client/DaprApiException.cs:40]: What's your suggestion? I kept the structure of these two types based on the existing places they are thrown and handled in Actors.

File: src/Dapr.Client/DaprClientGrpc.cs (hunk 6 above)
Comment (vinayada1) [reply src/Dapr.Client/DaprClientGrpc.cs:281]: ok...makes sense

File: test/Dapr.Client.Test/PublishEventApiTest.cs (hunk 5 above)
Comment (vinayada1) [reply test/Dapr.Client.Test/PublishEventApiTest.cs:90]: Yes, passing null as the value is the ut I am suggesting since that's the case we found was broken and also establishes the behavior of the API if someone is looking. I agree that writing tests that are more like json serializer tests is unnecessary.

File: src/Dapr.Client/DaprApiException.cs (hunk 7 above)
Comment (vinayada1) [reply src/Dapr.Client/DaprApiException.cs:40]: DaprActorException

**Review by vinayada1 — APPROVED** [review]

File: src/Dapr.Client/DaprApiException.cs (hunk 7 above)
Comment (rynowak) [reply src/Dapr.Client/DaprApiException.cs:40]: Discussed offline.

**Review by vinayada1 — DISMISSED** [review]

**Review by vinayada1 — APPROVED** [review]
